File: cpp/myLeetCode/Dynamic_Programming/Basic/343. Integer Break.cpp
// 343. Integer Break
// https://leetcode.com/problems/integer-break/

#include <iostream>
#include <vector>

int integerBreak(int n) {
    std::vector<int> dp(n + 1, 0);

    for (int i = 2; i <= n; i++) {
        for (int j = 1; j < i; j++) {
            dp[i] = std::max(dp[i], std::max(j * (i - j), j * dp[i - j]));
        }
    }

    return dp[n];
}

int main() {
    std::vector<int> testCases = {2, 10, 5};
    for (auto& testCase: testCases) {
        std::cout << "For n = " << testCase << ", the maximum product is: " << integerBreak(testCase) << std::endl;
    }
    
    return 0;
}

File: LeetCode/343-integer-break.py
def integerBreak(n: int) -> int:
    # 初始化 dp 数组，其中 dp[i] 表示将整数 i 拆分后得到的最大乘积
    dp = [0] * (n + 1)
    
    # 边界条件
    dp[1] = 1
    
    # 填充 dp 数组
    for i in range(2, n + 1):
        for j in range(1, i):
            dp[i] = max(dp[i], j * (i - j), j * dp[i - j])
    
    return dp[n]

# 示例调用
print(integerBreak(10))  # 输出应该是 36
